src/controller.js

```diff
@@ -0,0 +1,601 @@
+    }
+
+    if (foundKey !== key) {
+      throw new Error('Document contains incorrect Key');
+    }
+
+    if (typeof lookingfor === 'undefined') {
+      throw new Error('Document does not contain ' + property);
+    }
+
+    return lookingfor;
+  }
+  /**
+   * Obtain an UploadId from an Initiate Multipart Upload response body
+   */
+  __getUploadId(doc, bucket, key) {
+    return this.__getResponseProperty(doc, 'InitiateMultipartUploadResult', 'UploadId', bucket, key);
+  }
+
+  /**
+   * Obtain the ETag for the committed multipart upload
+   */
+  __getMultipartEtag(doc, bucket, key) {
+    return this.__getResponseProperty(doc, 'CompleteMultipartUploadResult', 'ETag', bucket, key);
+  }
+
+  // Return a list of 2-tuple's that are header key and value pairings for the
+  // headers to specify as the ACL for an object
+  __determinePermissionsHeaders(permissions) {
+    permissions = runSchema(permissions, schemas.permissions.required());
+    let {acl, read, write, readAcp, writeAcp, fullControl} = permissions;
+    if (acl) {
+      return [['x-amz-acl', permissions.acl]];
+    }
+
+    let perms = [];
+
+    if (read) {
+      perms.push(['x-amz-grant-read', read]);
+    }
+
+    if (write) {
+      perms.push(['x-amz-grant-write', write]);
+    }
+
+    if (readAcp) {
+      perms.push(['x-amz-grant-read-acp', readAcp]);
+    }
+
+    if (writeAcp) {
+      perms.push(['x-amz-grant-write-acp', writeAcp]);
+    }
+
+    if (fullControl) {
+      perms.push(['x-amz-grant-full-control', fullControl]);
+    }
+
+    return perms;
+  }
+
+  __s3hostname(bucket) {
+    let hostname = `${bucket}.${this.s3host}`;
+    if (this.s3port) {
+      hostname += ':' + this.s3port;
+    }
+    return hostname;
+  }
+ 
+  /**
+   * Initiate a Multipart upload and return the UploadIp that
+   * Amazon has assigned for this multipart upload
+   * http://docs.aws.amazon.com/AmazonS3/latest/API/mpUploadInitiate.html
+   */
+  async initiateMultipartUpload(opts) {
+
+    opts = runSchema(opts, Joi.object().keys({
+      bucket: schemas.bucket.required(),
+      key: schemas.key.required(),
+      sha256: schemas.sha256.required(),
+      size: schemas.mpSize.required(),
+      permissions: schemas.permissions,
+    }).optionalKeys('permissions'));
+
+    let {bucket, key, sha256, size, permissions} = opts;
+
+    if (size <= 0) {
+      // Each part must have a non-zero size
+      throw new Error('Objects must be more than 0 bytes');
+    } else if (size > 5 * 1024 * 1024 * 1024 * 1024) {
+      // The entire file must be lower than 5 TB
+      throw new Error('Object must total fewer than 5 TB'); 
+    }
+    let unsignedRequest = {
+      service: 's3',
+      region: this.region,
+      method: 'POST',
+      protocol: this.s3protocol,
+      hostname: this.__s3hostname(bucket),
+      path: `/${key}?uploads=`,
+      headers: {
+        'x-amz-meta-content-sha256': sha256,
+        'x-amz-meta-content-length': size,
```

Comment:
Ah, my thinking was that S3 would calculate it and that would be enough...

I think it'll be confusing to have two... Maybe we can just validate the content-length that s3 reports to the we got from the worker..

Hmm, just looked it up and would be an extra head request....

But when we think about it for a bit... Each of the part upload requests includes a size... And those sizes are signed so they must match... And we can't finish without all parts .. so content-length can't be sorter than what was declared in createArtifact...  Even with transport errors... It would have to be an S3, which I don't think we should harden against.